### ISSUE TITLE
Multiple outdoor units

### ISSUE THREAD

**Issue by sodabrew** [gonzojive/heatpump#24]
I installed two CX34 units. At a high level, I can see two options for talking to them:

1. Two RS-485 busses, each machine is ID 1 on its own bus, run two copies of cx34collector.
2. One RS-485 bus, change the ID of one of the machines (it's a setting on the touch screen controller), modify cx34collector to accept a list of bus IDs to collect from.

Other thoughts:
1. Probably should have separate database files per machine.
2. Cx34dash would be best with one web interface and either overlaid or side-by-side charts so any differences can be quickly observed. If there are two collector instances this means a list of grpc ports to talk to. 

Reading the Connect method in cx34.go, the slaveId variable seems bound up with the serial port instance. Would need to separate these so that one serial port instance is used by multiple modbus instances ... making sure frames aren't clobbered on the wire at the same time.

**Comment by gonzojive** [comment]
2 would be best. Some sort of configuration file that specifies what devices are available and what serial device should be used to connect to them sounds appropriate. The same config file could be used for the fan coil units, too. It'd be good to also assign a UUID to each device. AFAIK there is no serial number info available through modbus commands.

```protobuf
// config.proto

message Config {
  repeated HeatPump heat_pumps = 1;
  repeated FanCoilUnit fan_coil_units = 2;
}

message HeatPump {
  // Address of the device.
  ModbusServerSpec address = 1;
}

message FanCoilUnit {
  // Address of the device.
  ModbusServerSpec address = 1;
}

message ModbusServerSpec {
  // Modbus address of the device. 2 bits.
  uint32 server_id = 1;

  // Serial device used to connect to the server. Value of this ifeld is something like "/dev/ttyX"
  string serial_device= 2;
}
```

I plan to remove the [offensive terms](https://modbus.org/docs/Client-ServerPR-07-2020-final.docx.pdf) from the library at some point (replace slaveID with serverID).

**Comment by sodabrew** [comment]
+1. Just confirming that I understand: in the Modbus model, the device is the server and this software on the RPi is the client?

**Comment by gonzojive** [comment]
> +1. Just confirming that I understand: in the Modbus model, the device is the server and this software on the RPi is the client?

Yes, that's my understanding of the how it works.

`prototext` is the Go library for the text format, and it's decently readable. What's nice vs. JSON is there is more validation. https://marketplace.visualstudio.com/items?itemName=thesofakillers.vscode-pbtxt

Protos have a JSON encoding that would allow storing the config in JSON or YAML, but proto text format is simpler.

**Comment by sodabrew** [comment]
Agreed, protobuf-json format is rough. I'll take a look at proto-text.

**Comment by sodabrew** [comment]
Something that's not totally clear to me: can you run the fancoil client on the same RS-485 device as the cx34 client? Each sets up its own instance of `handler := modbus.NewRTUClientHandler(p.TTYDevice)` -- is there an underlying locking mechanism that prevents the two from clobbering each other on the same serial bus?

**Comment by gonzojive** [comment]
> Something that's not totally clear to me: can you run the fancoil client on the same RS-485 device as the cx34 client? Each sets up its own instance of `handler := modbus.NewRTUClientHandler(p.TTYDevice)` -- is there an underlying locking mechanism that prevents the two from clobbering each other on the same serial bus?

I'm not sure. I don't think there is any locking mechanism and suspect using the same device would be unsafe. Both Go processes would need to be able to read the same bytes from the same device to process traffic on the bus. Writes would also require locking.

I was a little paranoid myself and bought two USB RS485 dongles just in case... one for the fan coils and one for the heat pump.

--

Using a different modbus library would be fine with me if upstream isn't accepting patches.

**Comment by gonzojive** [comment]
It looks like serial port locking comes up and two processes will cause problems.
https://stackoverflow.com/questions/17980725/locking-linux-serial-port/17983721#17983721

One software solution is to update this code to use a single process for reading/writing from the serial device. Another is probably closing the serial device when not in use and protecting opens with a lock. I would just grab another USB to RS485 converter for now. The Waveshare ones seem pretty high quality and are ~$20.

**Comment by njmckenzie** [comment]
FYI that I got some really weird behavior using a single RS485 even with a single process (I think).  I'm using the logic from this project but putting it in OpenHab instead of Google.  I'm not 100% sure the openHab Modbus code is actually using a single process but it was easy rationale to stick with two RS-485 devices.  When I polled certain addresses (no discernible pattern) on the CX34, it would cause all the fan coils to start blinking gibberish and stop working.  The instant I deactivated the CX34 polling, they returned to normal.

**Comment by sodabrew** [comment]
What other devices were on your RS-485 bus? In theory, one bus could have many different components from different vendors; but I wonder if building automation folks know that in practice there are known issues?

It's been a couple of months since I had time to hack on this, but I think the path to using the same bus will be factoring out a "bus manager" from the other components of the system. The more I sketched my idea, the more I realized I was reinventing [Modbus over TCP](https://en.wikipedia.org/wiki/Modbus#Protocol_versions) -- that is, the "bus manager" would sit on the serial port and expose a TCP port and client programs would use Modbus over TCP. That seems to be where the industry is, definitely feel like the hobbyist playing catch-up :) 

This codebase mentions using Omron format. I wasn't able to find any descriptions on the web of how that interacts with Modbus over TCP.

**Comment by gonzojive** [comment]
I think the issue is that there are multiple processes using the same
serial device on your raspberry pi. They are likely reading or writing in
incompatible ways to the serial device.

Modbus itself can handle lots of devices as long as they are well behaving
and all using the same baud rate and other network parameters.

**Comment by sodabrew** [comment]
It's finally spring season, which means switching between heating and cooling modes every few days. Time to dust off my Pi and get this working! I tinkered with the collector but realized I wanted a more minimal starting point. So I started ripping down to the bare bones. I have a small command line output now. I haven't committed to my fork yet.

The mode and target temperatures are the things I need to be able to change remotely, so I'll work on adding write support to those registers from the command line. Here's the first light I have so far:

```
~/chilcmd $ ./chilcmd -tty /dev/ttyUSB0 -unit 2
Summary for CX34 unit 2:
  Mode: Cooling + DHW
  COP: 0.00 (stopped)
  Power: 0.00 Watts
  Outdoor Temp: 55.04 ºF
  Active Target Temp: 87.80 ºF
  Heating Target Temp: 87.80 ºF
  Inlet Temp: 55.76 ºF
  Outlet Temp: 55.40 ºF
  Pump Speed: 0.01 l/s
  Useful Heat Rate: 0.1163kg/s * -0.2°K * 4kJ/(kg * °K) = -93J/s = -0.09kW
```
